### ISSUE TITLE
Abort queries when replicas are OOMing

### ISSUE THREAD

**Issue by umanwizard** [MaterializeInc/materialize#20604]
### Feature request

If all replicas have OOMed since a query was emitted, we should abort that query.

We can do this by keeping track in the compute instance of the last time each replica has OOMed, and when the minimum across all those values changes, send a message (instance_id, least recent replica OOM) to the coordinator which will cause it to cancel all queries more recent than that time.

**Comment by ggevay** [comment]
> We can do this by keeping track in the compute instance of the last time each replica has OOMed, and when the minimum across all those values changes, send a message (instance_id, least recent replica OOM) to the coordinator which will cause it to cancel all queries more recent than that time.

I'm not sure I correctly understand this, but it seems to me that this wouldn't cancel the query that caused the OOM, because this would cancel queries that were issued after the OOM. This might still be useful though, to prevent more queries from piling up and exacerbating the problem.

Another (maybe complimentary) approach would be to abort all such ad-hoc queries that have been active through two OOMs in all replicas.

**Comment by umanwizard** [comment]
> it seems to me that this wouldn't cancel the query that caused the OOM, because this would cancel queries that were issued after the OOM

Yes, sorry. I meant to write the opposite of what I in fact wrote.
